File: app/src/main/java/com/teambrella/android/ui/home/HomeCardsFragment.java
package com.teambrella.android.ui.home;

import android.content.Context;
import android.content.res.Resources;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.constraint.ConstraintLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.ViewPager;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.bumptech.glide.RequestBuilder;
import com.bumptech.glide.load.resource.bitmap.CenterCrop;
import com.bumptech.glide.load.resource.bitmap.RoundedCorners;
import com.bumptech.glide.request.RequestOptions;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.teambrella.android.R;
import com.teambrella.android.api.TeambrellaModel;
import com.teambrella.android.api.model.json.JsonWrapper;
import com.teambrella.android.image.glide.GlideApp;
import com.teambrella.android.ui.IMainDataHost;
import com.teambrella.android.ui.base.ADataFragment;
import com.teambrella.android.ui.base.ITeambrellaDaggerActivity;
import com.teambrella.android.ui.base.dagger.IDaggerActivity;
import com.teambrella.android.ui.chat.ChatActivity;
import com.teambrella.android.util.AmountCurrencyUtil;

import io.reactivex.Notification;

/**
 * Home Cards Fragment.
 */
public class HomeCardsFragment extends ADataFragment<IMainDataHost> {


    private TextView mHeader;
    private TextView mSubHeader;
    private ViewPager mCardsPager;
    private CardAdapter mAdapter;
    private LinearLayout mPagerIndicator;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_home_cards, container, false);
        mHeader = view.findViewById(R.id.home_header);
        mCardsPager = view.findViewById(R.id.cards_pager);
        mPagerIndicator = view.findViewById(R.id.page_indicator);
        mSubHeader = view.findViewById(R.id.home_sub_header);
        mCardsPager.setPageMargin(20);
        return view;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof IDaggerActivity) {
            ((ITeambrellaDaggerActivity) context).getComponent().inject(this);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mAdapter = null;
    }

    @Override
    protected void onDataUpdated(Notification<JsonObject> notification) {
        if (notification.isOnNext()) {
            JsonWrapper response = new JsonWrapper(notification.getValue());
            JsonWrapper data = response.getObject(TeambrellaModel.ATTR_DATA);
            mHeader.setText(getString(R.string.welcome_user_format_string, data.getString(TeambrellaModel.ATTR_DATA_NAME).trim().split(" ")[0]));
            JsonArray cards = data.getJsonArray(TeambrellaModel.ATTR_DATA_CARDS);

            if (mSubHeader != null) {
                mSubHeader.setVisibility(View.VISIBLE);
            }

            if (mAdapter == null) {
                mCardsPager.setAdapter(mAdapter = new CardAdapter(cards));
                LayoutInflater inflater = LayoutInflater.from(getContext());
                for (int i = 0; i < cards.size(); i++) {
                    View view = inflater.inflate(R.layout.home_card_pager_indicator, mPagerIndicator, false);
                    view.setSelected(mCardsPager.getCurrentItem() == i);
                    mPagerIndicator.addView(view);
                }

                mCardsPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
                    @Override
                    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

                    }

                    @Override
                    public void onPageSelected(int position) {
                        for (int i = 0; i < mPagerIndicator.getChildCount(); i++) {
                            mPagerIndicator.getChildAt(i).setSelected(position == i);
                        }
                    }

                    @Override
                    public void onPageScrollStateChanged(int state) {
                        ((HomeFragment) getParentFragment()).setRefreshingEnable(state == ViewPager.SCROLL_STATE_IDLE);
                    }
                });
            } else {
                mAdapter.setData(cards);
            }
        }
    }


    public class CardAdapter extends FragmentStatePagerAdapter {

        private JsonArray mCards;

        CardAdapter(JsonArray cards) {
            super(HomeCardsFragment.this.getChildFragmentManager());
            mCards = cards;
        }

        @Override
        public Fragment getItem(int position) {
            return CardsFragment.getInstance(position, mTags);
        }

        @Override
        public int getCount() {
            return mCards.size();
        }

        public void setData(JsonArray cards) {
            mCards = cards;
            notifyDataSetChanged();
        }

        @Override
        public int getItemPosition(Object object) {
            return POSITION_UNCHANGED;
        }
    }


    public static final class CardsFragment extends ADataFragment<IMainDataHost> {

        private static final String EXTRA_POSITION = "position";
        ImageView icon;
        ImageView teammtePicture;
        TextView message;
        TextView unread;
        TextView amountWidget;
        TextView teamVote;
        TextView title;
        TextView subtitle;
        TextView leftTitle;
        View votingLabel;
        int mPosition;

        public static CardsFragment getInstance(int position, String[] tags) {
            CardsFragment fragment = new CardsFragment();
            Bundle args = new Bundle();
            args.putInt(EXTRA_POSITION, position);
            args.putStringArray(EXTRA_DATA_FRAGMENT_TAG, tags);
            fragment.setArguments(args);
            return fragment;
        }

        @Override
        public void onCreate(@Nullable Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            mPosition = getArguments().getInt(EXTRA_POSITION);
        }

        @NonNull
        @Override
        public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
            View view = inflater.inflate(R.layout.home_card_claim, container, false);
            icon = view.findViewById(R.id.icon);
            message = view.findViewById(R.id.message_text);
            unread = view.findViewById(R.id.unread);
            amountWidget = view.findViewById(R.id.amount_widget);
            teamVote = view.findViewById(R.id.team_vote);
            title = view.findViewById(R.id.title);
            subtitle = view.findViewById(R.id.subtitle);
            leftTitle = view.findViewById(R.id.left_title);
            votingLabel = view.findViewById(R.id.voting_label);
            teammtePicture = view.findViewById(R.id.teammate_picture);
            return view;
        }

        @Override
        protected void onDataUpdated(Notification<JsonObject> notification) {

            if (notification.isOnError()) {
                return;
            }

            JsonWrapper response = new JsonWrapper(notification.getValue());
            JsonWrapper data = response.getObject(TeambrellaModel.ATTR_DATA);
            JsonArray cards = data.getJsonArray(TeambrellaModel.ATTR_DATA_CARDS);
            JsonWrapper card = new JsonWrapper(cards.get(mPosition).getAsJsonObject());

            int itemType = card.getInt(TeambrellaModel.ATTR_DATA_ITEM_TYPE);

            leftTitle.setText(itemType == TeambrellaModel.FEED_ITEM_TEAMMATE ? R.string.coverage : R.string.claimed);


            RequestBuilder requestCreator = GlideApp.with(this).load(getImageLoader().getImageUrl(card.getString(TeambrellaModel.ATTR_DATA_SMALL_PHOTO_OR_AVATAR)));

            if (itemType == TeambrellaModel.FEED_ITEM_TEAMMATE
                    || itemType == TeambrellaModel.FEED_ITEM_TEAM_CHAT) {
                requestCreator = requestCreator.apply(RequestOptions.circleCropTransform());
            } else {
                Resources resources = getContext().getResources();
                requestCreator = requestCreator.apply(new RequestOptions()
                        .transforms(new CenterCrop(), new RoundedCorners(resources.getDimensionPixelOffset(R.dimen.rounded_corners_3dp))));
            }

            requestCreator.into(icon);

            message.setText(Html.fromHtml(card.getString(TeambrellaModel.ATTR_DATA_TEXT, "")));
            message.post(() -> message.setMaxLines(message.length() > 64 ? 2 : 1));

            int unreadCount = card.getInt(TeambrellaModel.ATTR_DATA_UNREAD_COUNT);

            unread.setVisibility(unreadCount > 0 ? View.VISIBLE : View.INVISIBLE);
            unread.setText(card.getString(TeambrellaModel.ATTR_DATA_UNREAD_COUNT));

            AmountCurrencyUtil.setAmount(amountWidget, card.getFloat(TeambrellaModel.ATTR_DATA_AMOUNT), mDataHost.getCurrency());
            if (itemType == TeambrellaModel.FEED_ITEM_TEAMMATE) {
                teamVote.setText(getString(R.string.risk_format_string, card.getFloat(TeambrellaModel.ATTR_DATA_TEAM_VOTE)));
            } else {
                teamVote.setText(Html.fromHtml(getString(R.string.home_team_vote_format_string, Math.round(card.getFloat(TeambrellaModel.ATTR_DATA_TEAM_VOTE) * 100))));
            }


            switch (itemType) {
                case TeambrellaModel.FEED_ITEM_CLAIM:
                    title.setText(card.getString(TeambrellaModel.ATTR_DATA_MODEL_OR_NAME));
                    break;
                case TeambrellaModel.FEED_ITEM_TEAM_CHAT:
                    title.setText(card.getString(TeambrellaModel.ATTR_DATA_CHAT_TITLE));
                    break;
                case TeambrellaModel.FEED_ITEM_TEAMMATE:
                    title.setText(card.getString(TeambrellaModel.ATTR_DATA_ITEM_USER_NAME));
                    break;
            }

            if (itemType == TeambrellaModel.FEED_ITEM_CLAIM) {
                subtitle.setText(card.getString(TeambrellaModel.ATTR_DATA_ITEM_USER_NAME));
                teammtePicture.setVisibility(View.VISIBLE);
                GlideApp.with(this).load(getImageLoader().getImageUrl(card.getString(TeambrellaModel.ATTR_DATA_ITEM_USER_AVATAR)))
                        .into(teammtePicture);
                ConstraintLayout.LayoutParams params = (ConstraintLayout.LayoutParams) subtitle.getLayoutParams();
                params.setMarginStart(getContext().getResources().getDimensionPixelOffset(R.dimen.margin_4));
                subtitle.setLayoutParams(params);
            } else {
                subtitle.setText(getContext().getString(R.string.object_format_string
                        , card.getString(TeambrellaModel.ATTR_DATA_MODEL_OR_NAME)
                        , card.getString(TeambrellaModel.ATTR_DATA_YEAR)
                ));
                teammtePicture.setVisibility(View.GONE);
                ConstraintLayout.LayoutParams params = (ConstraintLayout.LayoutParams) subtitle.getLayoutParams();
                params.setMarginStart(0);
                subtitle.setLayoutParams(params);
            }


            votingLabel.setVisibility(card.getBoolean(TeambrellaModel.ATTR_DATA_IS_VOTING, false) ? View.VISIBLE : View.GONE);

            View view = getView();
            if (view != null) {
                view.setOnClickListener(v -> {
                    Context context = getContext();
                    IMainDataHost dataHost = (IMainDataHost) getContext();
                    switch (itemType) {
                        case TeambrellaModel.FEED_ITEM_CLAIM:
                            dataHost.launchActivity(ChatActivity.getClaimChat(context
                                    , mDataHost.getTeamId()
                                    , card.getInt(TeambrellaModel.ATTR_DATA_ITEM_ID)
                                    , card.getString(TeambrellaModel.ATTR_DATA_MODEL_OR_NAME)
                                    , card.getString(TeambrellaModel.ATTR_DATA_SMALL_PHOTO_OR_AVATAR)
                                    , card.getString(TeambrellaModel.ATTR_DATA_TOPIC_ID)
                                    , mDataHost.getTeamAccessLevel()
                                    , card.getString(TeambrellaModel.ATTR_DATA_ITEM_DATE)));
                            break;
                        case TeambrellaModel.FEED_ITEM_TEAM_CHAT:
                            dataHost.launchActivity(ChatActivity.getFeedChat(context
                                    , card.getString(TeambrellaModel.ATTR_DATA_CHAT_TITLE)
                                    , card.getString(TeambrellaModel.ATTR_DATA_TOPIC_ID)
                                    , dataHost.getTeamId()
                                    , mDataHost.getTeamAccessLevel()));
                            break;
                        default:
                            dataHost.launchActivity(ChatActivity.getTeammateChat(context, mDataHost.getTeamId()
                                    , card.getString(TeambrellaModel.ATTR_DATA_ITEM_USER_ID)
                                    , null
                                    , card.getString(TeambrellaModel.ATTR_DATA_SMALL_PHOTO_OR_AVATAR)
                                    , card.getString(TeambrellaModel.ATTR_DATA_TOPIC_ID)
                                    , mDataHost.getTeamAccessLevel()));
                            break;
                    }
                });
            }
        }
    }

}

File: app/src/main/java/com/teambrella/android/ui/home/HomeCardsFragment.kt
package com.teambrella.android.ui.home

import android.os.Bundle
import android.support.constraint.ConstraintLayout
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentStatePagerAdapter
import android.support.v4.view.PagerAdapter
import android.support.v4.view.ViewPager
import android.text.Html
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.google.gson.JsonObject
import com.teambrella.android.R
import com.teambrella.android.api.*
import com.teambrella.android.ui.IMainDataHost
import com.teambrella.android.ui.base.ADataFragment
import com.teambrella.android.ui.base.AKDataFragment
import com.teambrella.android.ui.chat.ChatActivity
import com.teambrella.android.ui.util.setAvatar
import com.teambrella.android.ui.util.setImage
import com.teambrella.android.ui.util.setUnreadCount
import com.teambrella.android.util.AmountCurrencyUtil
import io.reactivex.Notification

class HomeCardsFragment : AKDataFragment<IMainDataHost>() {

    private val header: TextView? by ViewHolder(R.id.home_header)
    private val subHeader: TextView? by ViewHolder(R.id.home_sub_header)
    private val cardsPager: ViewPager? by ViewHolder(R.id.cards_pager)
    private var adapter: CardsAdapter? = null
    private val pagerIndicator: LinearLayout? by ViewHolder(R.id.page_indicator)

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater.inflate(R.layout.fragment_home_cards, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        cardsPager?.pageMargin = 20
    }

    override fun onDataUpdated(notification: Notification<JsonObject>?) {
        val data = notification?.takeIf { it.isOnNext }?.value?.data
        data?.let { _data ->
            header?.text = getString(R.string.welcome_user_format_string, _data.name?.trim()?.split(" ".toRegex())!![0])
            subHeader?.visibility = View.VISIBLE

            if (adapter == null) {

                adapter = CardsAdapter(_data.cards?.size() ?: 0)
                cardsPager?.adapter = adapter

                val inflater = LayoutInflater.from(context)
                _data.cards?.forEachIndexed { index, _ ->
                    val view = inflater.inflate(R.layout.home_card_pager_indicator, pagerIndicator, false)
                    view.isSelected = cardsPager?.currentItem ?: 0 == index
                    pagerIndicator?.addView(view)
                }

                cardsPager?.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
                    override fun onPageScrollStateChanged(state: Int) {
                        (parentFragment as KHomeFragment).refreshingEnabled = state == ViewPager.SCROLL_STATE_IDLE
                    }

                    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) = Unit

                    override fun onPageSelected(position: Int) {
                        for (i in 0..(pagerIndicator?.childCount ?: 0)) {
                            pagerIndicator?.getChildAt(i)?.isSelected =
                                    (i == position)
                        }
                    }
                })
            } else {
                adapter?.notifyDataSetChanged()
            }
        }
    }


    override fun onDestroyView() {
        super.onDestroyView()
        adapter = null
    }

    inner class CardsAdapter(val size: Int) : FragmentStatePagerAdapter(childFragmentManager) {
        override fun getItem(position: Int): Fragment = createCardFragment(position, mTags)
        override fun getCount() = size
        override fun getItemPosition(`object`: Any) = PagerAdapter.POSITION_UNCHANGED
    }
}

private const val EXTRA_POSITION = "position"

private fun createCardFragment(position: Int, tags: Array<String>): KCardFragment {
    return KCardFragment().apply {
        arguments = Bundle().apply {
            putInt(EXTRA_POSITION, position)
            putStringArray(ADataFragment.EXTRA_DATA_FRAGMENT_TAG, tags)
        }
    }
}

class KCardFragment : AKDataFragment<IMainDataHost>() {

    private val icon: ImageView? by ViewHolder(R.id.icon)
    private val teammatePicture: ImageView? by ViewHolder(R.id.teammate_picture)
    private val unread: TextView? by ViewHolder(R.id.unread)
    private val amountWidget: TextView? by ViewHolder(R.id.amount_widget)
    private val teamVote: TextView? by ViewHolder(R.id.team_vote)
    private val title: TextView? by ViewHolder(R.id.title)
    private val subtitle: TextView? by ViewHolder(R.id.subtitle)
    private val leftTile: TextView? by ViewHolder(R.id.left_title)
    private val votingLabel: TextView? by ViewHolder(R.id.voting_label)
    private val message: TextView? by ViewHolder(R.id.message_text)
    private var position: Int = 0


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        position = arguments?.getInt(EXTRA_POSITION, 0) ?: 0
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater.inflate(R.layout.home_card_claim, container, false)


    override fun onDataUpdated(notification: Notification<JsonObject>?) {
        val card = notification?.takeIf { it.isOnNext }?.value?.data?.cards!![position]?.asJsonObject
        card?.let { _card ->

            when (_card.itemType) {
                TeambrellaModel.FEED_ITEM_TEAMMATE -> {
                    leftTile?.setText(R.string.coverage)
                    icon?.setAvatar(imageLoader.getImageUrl(_card.smallPhotoOrAvatar))
                    title?.text = _card.itemUserName
                    teamVote?.text = getString(R.string.risk_format_string, _card.teamVote ?: 0f);
                    subtitle?.text = getString(R.string.object_format_string, _card.modelOrName, _card.year)
                    teammatePicture?.visibility = View.GONE
                    subtitle?.layoutParams = (subtitle?.layoutParams as ConstraintLayout.LayoutParams).apply {
                        marginStart = 0
                    }
                }
                TeambrellaModel.FEED_ITEM_CLAIM -> {
                    leftTile?.setText(R.string.claimed)
                    icon?.setImage(imageLoader.getImageUrl(_card.smallPhotoOrAvatar), R.dimen.rounded_corners_3dp)
                    title?.text = _card.modelOrName
                    teamVote?.text = Html.fromHtml(getString(R.string.home_team_vote_format_string, Math.round((_card.teamVote
                            ?: 0f) * 100)));
                    subtitle?.text = _card.itemUserName
                    teammatePicture?.visibility = View.VISIBLE
                    teammatePicture?.setAvatar(imageLoader.getImageUrl(_card.itemUserAvatar))
                    subtitle?.layoutParams = (subtitle?.layoutParams as ConstraintLayout.LayoutParams).apply {
                        marginStart = resources.getDimensionPixelOffset(R.dimen.margin_4)
                    }
                }
            }

            message?.text = Html.fromHtml(_card.text)
            message?.post({ message?.maxLines = if ((message?.length() ?: 0) > 64) 2 else 1 });


            unread?.setUnreadCount(_card.unreadCount ?: 0)

            AmountCurrencyUtil.setAmount(amountWidget, _card.amount ?: 0f, mDataHost.currency);

            votingLabel?.visibility = if (_card.isVoting == true) View.VISIBLE else View.GONE;

            view?.setOnClickListener {
                when (_card.itemType) {
                    TeambrellaModel.FEED_ITEM_CLAIM -> {
                        startActivity(ChatActivity.getClaimChat(context
                                , mDataHost.teamId
                                , _card.itemIdInt ?: 0
                                , _card.modelOrName
                                , _card.smallPhotoOrAvatar
                                , _card.topicId
                                , mDataHost.teamAccessLevel
                                , _card.itemDate))
                    }
                    else -> {
                        startActivity(ChatActivity.getTeammateChat(context, mDataHost.teamId
                                , _card.itemUserId
                                , _card.itemUserName
                                , _card.smallPhotoOrAvatar
                                , _card.topicId
                                , mDataHost.teamAccessLevel))
                    }
                }
            }
        }
    }
}